File: java/rocksjni/transaction_db.cc
// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree. An additional grant
// of patent rights can be found in the PATENTS file in the same directory.
//
// This file implements the "bridge" between Java and C++
// for rocksdb::TransactionDB.

#include <jni.h>

#include "include/org_rocksdb_TransactionDB.h"

#include "rocksdb/options.h"
#include "rocksdb/utilities/transaction.h"
#include "rocksdb/utilities/transaction_db.h"

#include "rocksjni/portal.h"

/*
 * Class:     org_rocksdb_TransactionDB
 * Method:    open
 * Signature: (JJLjava/lang/String;)J
 */
jlong Java_org_rocksdb_TransactionDB_open__JJLjava_lang_String_2(JNIEnv* env,
                                                                 jclass jcls, jlong joptions_handle, jlong jtxn_db_options_handle,
                                                                 jstring jdb_path) {
    auto* options = reinterpret_cast<rocksdb::Options*>(joptions_handle);
    auto* txn_db_options =
            reinterpret_cast<rocksdb::TransactionDBOptions*>(jtxn_db_options_handle);
    rocksdb::TransactionDB* tdb = nullptr;
    const char* db_path = env->GetStringUTFChars(jdb_path, nullptr);
    rocksdb::Status s =
            rocksdb::TransactionDB::Open(*options, *txn_db_options, db_path, &tdb);
    env->ReleaseStringUTFChars(jdb_path, db_path);

    if (s.ok()) {
        return reinterpret_cast<jlong>(tdb);
    } else {
        rocksdb::RocksDBExceptionJni::ThrowNew(env, s);
        return 0;
    }
}

/*
 * Class:     org_rocksdb_TransactionDB
 * Method:    open
 * Signature: (JJLjava/lang/String;[[B[J)[J
 */
jlongArray Java_org_rocksdb_TransactionDB_open__JJLjava_lang_String_2_3_3B_3J(
        JNIEnv* env, jclass jcls, jlong jdb_options_handle,
        jlong jtxn_db_options_handle, jstring jdb_path,
        jobjectArray jcolumn_names,
        jlongArray jcolumn_options_handles) {
    auto* db_options = reinterpret_cast<rocksdb::DBOptions*>(jdb_options_handle);
    auto* txn_db_options =
            reinterpret_cast<rocksdb::TransactionDBOptions*>(jtxn_db_options_handle);
    const char* db_path = env->GetStringUTFChars(jdb_path, nullptr);

    std::vector<rocksdb::ColumnFamilyDescriptor> column_families;

    jsize len_cols = env->GetArrayLength(jcolumn_names);
    jlong* jco = env->GetLongArrayElements(jcolumn_options_handles, nullptr);
    for (int i = 0; i < len_cols; i++) {
        jobject jcn = env->GetObjectArrayElement(jcolumn_names, i);
        jbyteArray jcn_ba = reinterpret_cast<jbyteArray>(jcn);
        jbyte* jcf_name = env->GetByteArrayElements(jcn_ba, nullptr);

        const int jcf_name_len = env->GetArrayLength(jcn_ba);
        std::string cf_name(reinterpret_cast<char *>(jcf_name), jcf_name_len);
        rocksdb::ColumnFamilyOptions* cf_options =
                reinterpret_cast<rocksdb::ColumnFamilyOptions*>(jco[i]);
        column_families.push_back(
                rocksdb::ColumnFamilyDescriptor(cf_name, *cf_options));

        env->ReleaseByteArrayElements(jcn_ba, jcf_name, JNI_ABORT);
        env->DeleteLocalRef(jcn);
    }
    env->ReleaseLongArrayElements(jcolumn_options_handles, jco, JNI_ABORT);

    std::vector<rocksdb::ColumnFamilyHandle*> handles;
    rocksdb::TransactionDB* tdb = nullptr;
    rocksdb::Status s = rocksdb::TransactionDB::Open(*db_options, *txn_db_options,
                                                     db_path, column_families, &handles, &tdb);

    // check if open operation was successful
    if (s.ok()) {
        jsize resultsLen = 1 + len_cols;  // db handle + column family handles
        std::unique_ptr<jlong[]> results =
                std::unique_ptr<jlong[]>(new jlong[resultsLen]);
        results[0] = reinterpret_cast<jlong>(tdb);
        for (int i = 1; i <= len_cols; i++) {
            results[i] = reinterpret_cast<jlong>(handles[i - 1]);
        }

        jlongArray jresults = env->NewLongArray(resultsLen);
        env->SetLongArrayRegion(jresults, 0, resultsLen, results.get());
        return jresults;
    } else {
        rocksdb::RocksDBExceptionJni::ThrowNew(env, s);
        return nullptr;
    }
}

/*
 * Class:     org_rocksdb_TransactionDB
 * Method:    beginTransaction
 * Signature: (JJ)J
 */
jlong Java_org_rocksdb_TransactionDB_beginTransaction__JJ(JNIEnv* env,
                                                          jobject jobj, jlong jhandle, jlong jwrite_options_handle) {
    auto* txn_db = reinterpret_cast<rocksdb::TransactionDB*>(jhandle);
    auto* write_options =
            reinterpret_cast<rocksdb::WriteOptions*>(jwrite_options_handle);
    rocksdb::Transaction* txn = txn_db->BeginTransaction(*write_options);
    return reinterpret_cast<jlong>(txn);
}

/*
 * Class:     org_rocksdb_TransactionDB
 * Method:    beginTransaction
 * Signature: (JJJ)J
 */
jlong Java_org_rocksdb_TransactionDB_beginTransaction__JJJ(JNIEnv* env,
                                                           jobject jobj, jlong jhandle, jlong jwrite_options_handle,
                                                           jlong jtxn_options_handle) {
    auto* txn_db = reinterpret_cast<rocksdb::TransactionDB*>(jhandle);
    auto* write_options =
            reinterpret_cast<rocksdb::WriteOptions*>(jwrite_options_handle);
    auto* txn_options =
            reinterpret_cast<rocksdb::TransactionOptions*>(jtxn_options_handle);
    rocksdb::Transaction* txn =
            txn_db->BeginTransaction(*write_options, *txn_options);
    return reinterpret_cast<jlong>(txn);
}

/*
 * Class:     org_rocksdb_TransactionDB
 * Method:    beginTransaction_withOld
 * Signature: (JJJ)J
 */
jlong Java_org_rocksdb_TransactionDB_beginTransaction_1withOld__JJJ(
        JNIEnv* env, jobject jobj, jlong jhandle, jlong jwrite_options_handle,
        jlong jold_txn_handle) {
    auto* txn_db = reinterpret_cast<rocksdb::TransactionDB*>(jhandle);
    auto* write_options =
            reinterpret_cast<rocksdb::WriteOptions*>(jwrite_options_handle);
    auto* old_txn = reinterpret_cast<rocksdb::Transaction*>(jold_txn_handle);
    rocksdb::TransactionOptions txn_options;
    rocksdb::Transaction* txn =
            txn_db->BeginTransaction(*write_options, txn_options, old_txn);

    // RocksJava relies on the assumption that
    // we do not allocate a new Transaction object
    // when providing an old_txn
    assert(txn == old_txn);

    return reinterpret_cast<jlong>(txn);
}

/*
 * Class:     org_rocksdb_TransactionDB
 * Method:    beginTransaction_withOld
 * Signature: (JJJJ)J
 */
jlong Java_org_rocksdb_TransactionDB_beginTransaction_1withOld__JJJJ(
        JNIEnv* env, jobject jobj, jlong jhandle, jlong jwrite_options_handle,
        jlong jtxn_options_handle, jlong jold_txn_handle) {
    auto* txn_db = reinterpret_cast<rocksdb::TransactionDB*>(jhandle);
    auto* write_options =
            reinterpret_cast<rocksdb::WriteOptions*>(jwrite_options_handle);
    auto* txn_options =
            reinterpret_cast<rocksdb::TransactionOptions*>(jtxn_options_handle);
    auto* old_txn = reinterpret_cast<rocksdb::Transaction*>(jold_txn_handle);
    rocksdb::Transaction* txn = txn_db->BeginTransaction(*write_options,
                                                         *txn_options, old_txn);

    // RocksJava relies on the assumption that
    // we do not allocate a new Transaction object
    // when providing an old_txn
    assert(txn == old_txn);

    return reinterpret_cast<jlong>(txn);
}

/*
 * Class:     org_rocksdb_TransactionDB
 * Method:    getTransactionByName
 * Signature: (JLjava/lang/String;)J
 */
jlong Java_org_rocksdb_TransactionDB_getTransactionByName(JNIEnv* env,
                                                          jobject jobj, jlong jhandle, jstring jname) {
    auto* txn_db = reinterpret_cast<rocksdb::TransactionDB*>(jhandle);
    const char* name = env->GetStringUTFChars(jname, nullptr);
    rocksdb::Transaction* txn = txn_db->GetTransactionByName(name);
    env->ReleaseStringUTFChars(jname, name);
    return reinterpret_cast<jlong>(txn);
}

/*
 * Class:     org_rocksdb_TransactionDB
 * Method:    getAllPreparedTransactions
 * Signature: (J)[J
 */
jlongArray Java_org_rocksdb_TransactionDB_getAllPreparedTransactions(
        JNIEnv* env, jobject jobj, jlong jhandle) {
    auto* txn_db = reinterpret_cast<rocksdb::TransactionDB*>(jhandle);
    std::vector<rocksdb::Transaction*> txns;
    txn_db->GetAllPreparedTransactions(&txns);

    const size_t size = txns.size();
    assert(size < UINT32_MAX);  // does it fit in a jint?

    jsize len = static_cast<jsize>(size);
    jlong* tmp = new jlong[len];
    for (jsize i = 0; i < len; ++i) {
        tmp[i] = reinterpret_cast<jlong>(txns[i]);
    }

    jlongArray jtxns = env->NewLongArray(len);
    env->SetLongArrayRegion(jtxns, 0, len, tmp);
    delete [] tmp;

    return jtxns;
}

/*
 * Class:     org_rocksdb_TransactionDB
 * Method:    disposeInternal
 * Signature: (J)V
 */
void Java_org_rocksdb_TransactionDB_disposeInternal(JNIEnv* env, jobject jobj,
                                                    jlong jhandle) {
    delete reinterpret_cast<rocksdb::TransactionDB*>(jhandle);
}

File: java/src/main/java/org/rocksdb/TransactionDB.java
// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree. An additional grant
// of patent rights can be found in the PATENTS file in the same directory.

package org.rocksdb;

import java.util.ArrayList;
import java.util.List;

/**
 * Database with Transaction support
 */
public class TransactionDB extends RocksDB
    implements TransactionalDB<TransactionOptions> {

  private TransactionDBOptions transactionDbOptions_;

  /**
   * Private constructor.
   *
   * @param nativeHandle The native handle of the C++ TransactionDB object
   */
  private TransactionDB(final long nativeHandle) {
    super(nativeHandle);
  }

  /**
   * Open a TransactionDB, similar to {@link RocksDB#open(Options, String)}
   */
  public static TransactionDB open(final Options options,
      final TransactionDBOptions transactionDbOptions, final String path)
      throws RocksDBException {
    final TransactionDB tdb = new TransactionDB(open(options.nativeHandle_,
        transactionDbOptions.nativeHandle_, path));

    // when non-default Options is used, keeping an Options reference
    // in RocksDB can prevent Java to GC during the life-time of
    // the currently-created RocksDB.
    tdb.storeOptionsInstance(options);
    tdb.storeTransactionDbOptions(transactionDbOptions);

    return tdb;
  }

  /**
   * Open a TransactionDB, similar to
   * {@link RocksDB#open(DBOptions, String, List, List)}
   */
  public static TransactionDB open(final DBOptions dbOptions,
      final TransactionDBOptions transactionDbOptions,
      final String path,
      final List<ColumnFamilyDescriptor> columnFamilyDescriptors,
      final List<ColumnFamilyHandle> columnFamilyHandles)
      throws RocksDBException {

    final byte[][] cfNames = new byte[columnFamilyDescriptors.size()][];
    final long[] cfOptionHandles = new long[columnFamilyDescriptors.size()];
    for (int i = 0; i < columnFamilyDescriptors.size(); i++) {
      final ColumnFamilyDescriptor cfDescriptor = columnFamilyDescriptors
          .get(i);
      cfNames[i] = cfDescriptor.columnFamilyName();
      cfOptionHandles[i] = cfDescriptor.columnFamilyOptions().nativeHandle_;
    }

    final long[] handles = open(dbOptions.nativeHandle_,
        transactionDbOptions.nativeHandle_, path, cfNames, cfOptionHandles);
    final TransactionDB tdb = new TransactionDB(handles[0]);

    // when non-default Options is used, keeping an Options reference
    // in RocksDB can prevent Java to GC during the life-time of
    // the currently-created RocksDB.
    tdb.storeOptionsInstance(dbOptions);
    tdb.storeTransactionDbOptions(transactionDbOptions);

    for (int i = 1; i < handles.length; i++) {
      columnFamilyHandles.add(new ColumnFamilyHandle(tdb, handles[i]));
    }

    return tdb;
  }

  @Override
  public Transaction beginTransaction(final WriteOptions writeOptions) {
    return new Transaction(this, beginTransaction(nativeHandle_,
        writeOptions.nativeHandle_));
  }

  @Override
  public Transaction beginTransaction(final WriteOptions writeOptions,
      final TransactionOptions transactionOptions) {
    return new Transaction(this, beginTransaction(nativeHandle_,
        writeOptions.nativeHandle_, transactionOptions.nativeHandle_));
  }

  // TODO(AR) consider having beingTransaction(... oldTransaction) set a
  // reference count inside Transaction, so that we can always call
  // Transaction#close but the object is only disposed when there are as many
  // closes as beginTransaction. Makes the try-with-resources paradigm easier for
  // java developers

  @Override
  public Transaction beginTransaction(final WriteOptions writeOptions,
      final Transaction oldTransaction) {
    final long jtxnHandle = beginTransaction_withOld(nativeHandle_,
        writeOptions.nativeHandle_, oldTransaction.nativeHandle_);

    // RocksJava relies on the assumption that
    // we do not allocate a new Transaction object
    // when providing an old_txn
    assert(jtxnHandle == oldTransaction.nativeHandle_);

    return oldTransaction;
  }

  @Override
  public Transaction beginTransaction(final WriteOptions writeOptions,
      final TransactionOptions transactionOptions,
      final Transaction oldTransaction) {
    final long jtxn_handle = beginTransaction_withOld(nativeHandle_,
        writeOptions.nativeHandle_, transactionOptions.nativeHandle_,
        oldTransaction.nativeHandle_);

    // RocksJava relies on the assumption that
    // we do not allocate a new Transaction object
    // when providing an old_txn
    assert(jtxn_handle == oldTransaction.nativeHandle_);

    return oldTransaction;
  }

  public Transaction getTransactionByName(final String transactionName) {
    final long jtxnHandle = getTransactionByName(nativeHandle_, transactionName);
    if(jtxnHandle == 0) {
      return null;
    }

    final Transaction txn = new Transaction(this, jtxnHandle);

    // this instance doesn't own the underlying C++ object
    txn.disOwnNativeHandle();

    return txn;
  }

  public List<Transaction> getAllPreparedTransactions() {
    final long[] jtxnHandles = getAllPreparedTransactions(nativeHandle_);

    final List<Transaction> txns = new ArrayList<>();
    for(final long jtxnHandle : jtxnHandles) {
      final Transaction txn = new Transaction(this, jtxnHandle);

      // this instance doesn't own the underlying C++ object
      txn.disOwnNativeHandle();

      txns.add(txn);
    }
    return txns;
  }

  private void storeTransactionDbOptions(
      final TransactionDBOptions transactionDbOptions) {
    this.transactionDbOptions_ = transactionDbOptions;
  }

  private static native long open(final long optionsHandle,
      final long transactionDbOptionsHandle, final String path)
      throws RocksDBException;
  private static native long[] open(final long dbOptionsHandle,
      final long transactionDbOptionsHandle, final String path,
      final byte[][] columnFamilyNames, final long[] columnFamilyOptions);
  private native long beginTransaction(final long handle,
      final long writeOptionsHandle);
  private native long beginTransaction(final long handle,
      final long writeOptionsHandle, final long transactionOptionsHandle);
  private native long beginTransaction_withOld(final long handle,
      final long writeOptionsHandle, final long oldTransactionHandle);
  private native long beginTransaction_withOld(final long handle,
      final long writeOptionsHandle, final long transactionOptionsHandle,
      final long oldTransactionHandle);
  private native long getTransactionByName(final long handle,
      final String name);
  private native long[] getAllPreparedTransactions(final long handle);
  @Override protected final native void disposeInternal(final long handle);
}